src/api/controller/front/index.js

```diff
@@ -1,33 +1,164 @@
+import koaWebpack from 'koa-webpack';
+import fs from 'fs';
 
-import dev from './dev';
-import prod from './prod';
+import rootReducer from '../../../app/js/public/reducers';
+import sagas from '../../../app/js/public/sagas';
+import configureStoreServer from '../../../app/js/configureStoreServer';
+import routes from '../../../app/js/public/routes';
+import phrasesForEn from '../../../app/js/i18n/translations/en';
+import webpackConfig from '../../../app/webpack.config.babel';
 
-const app = new Koa();
+const indexHtml = fs
+    .readFileSync(path.resolve(__dirname, '../../../app/custom/index.html'))
+    .toString();
 
-const middleware = process.env.NODE_ENV === 'development' ? dev() : prod();
+const initialState = {
+    fields: {
+        loading: false,
+        isSaving: false,
+        byName: {},
+        allValid: true,
+        list: [],
+        invalidFields: [],
+        editedFieldName: undefined,
+        editedValueFieldName: null,
+        configuredFieldName: null,
+        published: true,
+    },
+    polyglot: {
+        locale: 'en',
+        phrases: phrasesForEn,
+    },
+};
 
-app.use(async (ctx, next) => {
-    const uri = ctx.path;
+const renderFullPage = (html, preloadedState, helmet) =>
+    indexHtml
+        .replace(
+            '<div id="root"></div>',
+            `<div id="root"><div>${html}</div></div>`,
+        )
+        .replace(/<title>.*?<\/title>/, helmet.title.toString())
+        .replace(
+            '</head>',
+            `${helmet.meta.toString()}
+            ${helmet.link.toString()}
+            </head>`,
+        )
+        .replace(
+            '</body>',
+            `<script>window.__PRELOADED_STATE__ = ${JSON.stringify(
+                preloadedState,
+            ).replace(/</g, '\\u003c')}</script>
+            <script src="/index.js"></script>
+            </body>`,
+        );
 
-    if (uri.startsWith('/uid:/')
-        || uri.startsWith('/ark:/')
-        || uri.startsWith('/login')
-        || uri.startsWith('/home')
-        || uri.startsWith('/resource')) {
-        // Override the path so that webpack serves the application correctly
-        ctx.path = '/';
+const handleRender = async (ctx, next) => {
+    const { url, headers } = ctx.request;
+    if (url === '/__webpack_hmr') {
+        await next();
+        return;
     }
-
-    if (uri.startsWith('/admin/')) {
-        // Override the path so that webpack serves the application correctly
-        ctx.path = '/admin';
+    if (url === '/') {
+        ctx.redirect('/home');
+        return;
     }
 
-    await next();
-});
+    const muiTheme = getMuiTheme(
+        {},
+        {
+            userAgent: headers['user-agent'],
+        },
+    );
+    const history = createMemoryHistory(url);
+
+    ctx.body = await new Promise((resolve, reject) => {
+        match(
+            { history, routes, location: url },
+            (error, redirect, renderProps) => {
+                const store = configureStoreServer(
```

Comment:
Can't you export that block to make it testable, and reduce the size of the current file ?